### PR TITLE
Refactor/eslint

### PR DESCRIPTION
**Pull request by zorkow**
Linting:

* Adds `eslint`
* Updates packages to latest version
* Automatic linting reduces from `>3000` to `~1200` problems, introducing empty jsdoc templates in the process

remaining issues 531 problems.

### CODE REVIEW

**Review by dpvc — APPROVED** [review]
This looks great!  I am particularly pleased with the `let` to `const` changes, as I thought I would have to go through them all by hand for that.  Glad that this was automated (I hope).  Just a few questions in the comments.

Also, someplace we use `=` at the end of optional arguments with defaults and `?` for  optional ones in the `@param` comments.  Should that be enforced, now that things are being normalized?

Finally, are `@override` comments being added where needed automatically?  I know that I have added `@param` comments in places that they shouldn't be.  Do we want to clean that up, too?

File: ts/core/MmlTree/MmlNode.ts [hunk 1]
```diff
@@ -1461,19 +1463,19 @@ export class XMLNode extends AbstractMmlEmptyNode {
...
 
   /**
    * @param {object} xml  The XML content to be saved
    * @param {DOMAdaptor} adaptor DOM adaptor for the content
-   * @return {XMLNode}  The XML node (for chaining of method calls)
+   * @returns {XMLNode}  The XML node (for chaining of method calls)
    */
   public setXML(
-    xml: Object,
+    xml: Object, // eslint-disable-line
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/core/MmlTree/MmlNode.ts:1478]: This leads to an error
```
Argument of type 'N' is not assignable to parameter of type 'object'
```
So I left it for now.

File: ts/a11y/speech/SpeechUtil.ts [hunk 2]
```diff
@@ -101,11 +101,12 @@ function recurseSsml(
...
           instr.push({ mark: element.getAttribute('name') });
           break;
-        case 'say-as':
-          let txt = element.textContent;
+        case 'say-as': {
+          const txt = element.textContent;
           instr.push(Object.assign({ text: txt, character: true }, prosody));
           text.push(txt);
           break;
+        }
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/speech/SpeechUtil.ts:109]: Neither did I, but the linter was very adamant about it. 
Apparently `switch` statements do not have implicit scopes, so `const txt` is visible in the entire statement (a bit like a `var`).

File: ts/a11y/assistive-mml.ts [hunk 3]
```diff
@@ -270,14 +270,16 @@ export function AssistiveMmlMathDocumentMixin<
...
+     * @returns {string}        The serialization of the node
      */
     public toMML(node: MmlNode): string {
       return this.visitor.visitTree(node);
     }
 
     /**
      * Add assistive MathML to the MathItems in this MathDocument
+     *
+     * @returns The  assistive mml document.
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/assistive-mml.ts:282]: Indeed, there is a number of those missing. I had switched off the checks `jsdoc/require-param-type` and `jsdoc/require-returns-type`. However, I'll try to add some more.

File: ts/a11y/explorer/KeyExplorer.ts [hunk 4]
```diff
@@ -427,8 +440,16 @@ export class SpeechExplorer
...
-   * @constructor
-   * @extends {AbstractKeyExplorer}
+   * @param document The accessible math document.
+   * @param pool The explorer pool.
+   * @param region The speech region for the explorer.
+   * @param node The node the explorer is assigned to.
+   * @param brailleRegion The braille region.
+   * @param magnifyRegion The magnification region.
+   * @param _mml The internal math node.
+   * @param item The math item.
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/explorer/KeyExplorer.ts:450]: ditto.

File: ts/a11y/explorer/KeyExplorer.ts [hunk 5]
```diff
@@ -652,7 +677,7 @@ export class SpeechExplorer
   }
 
   /**
-   * @return The semantic node that is currently focused.
+   * @returns The semantic node that is currently focused.
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/explorer/KeyExplorer.ts:680]: Same her.

[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/explorer/KeyExplorer.ts:680]: No need to flag them. That's why we have the linter!

File: ts/components/package.ts [hunk 6]
```diff
@@ -27,7 +27,7 @@ import { CONFIG, Loader } from './loader.js';
 /*
  * The browser document (for creating scripts to load components)
  */
-declare var document: Document;
+declare let document: Document;
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/components/package.ts:30]: That was automatically changed by the linter, I assume to `let` as there is no assignment, so it probably assumes it's assigned elsewhere. But it seems to work with `const`.

File: ts/input/tex/StackItem.ts [hunk 7]
```diff
@@ -379,8 +400,10 @@ export abstract class BaseItem extends MmlStack implements StackItem {
   private _properties: PropList = {};
 
   /**
-   * @constructor
-   * @extends {MmlStack}
+   * @param factory
+   * @param {...any} nodes
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/input/tex/StackItem.ts:404]: We can actually give it a proper type.

File: ts/core/MmlTree/MmlNode.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpvc) [reply ts/core/MmlTree/MmlNode.ts:1478]: Looking into `XMLNode` a bit more, I wonder if it should not be `XMLNode<N, T, D>`, and then this `Object` could be `N` (and the `document` that currently uses `<any, any, any>` could use `<N, T, D>`.

File: ts/output/chtml/Wrappers/mfenced.ts [hunk 8]
```diff
@@ -117,9 +117,9 @@ export const ChtmlMfenced = (function <N, T, D>(): ChtmlMfencedClass<N, T, D> {
     ChtmlMfencedClass<N, T, D>
   >(ChtmlWrapper);
 
-  // Avoid message about base constructors not having the same type
-  //   (they should both be ChtmlWrapper<N, T, D>, but are thought of as different by typescript)
-  // @ts-expect-error
+  // @ts-expect-error Avoid message about base constructors not having the same
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/output/chtml/Wrappers/mfenced.ts:120]: Testing with v5.6.2 it is still there. 
However, I get an extra type issue in the `MathtoolsTags` so I leave the TS version for the time being. We can update in a separate PR.

File: ts/util/asyncLoad/system.ts [hunk 9]
```diff
@@ -16,15 +16,15 @@
...
  *
  * @author dpvc@mathjax.org (Davide Cervone)
  */
 
 import { mathjax } from '../../mathjax.js';
 
-declare var System: { import: (name: string, url?: string) => any };
-declare var __dirname: string;
+declare let System: { import: (name: string, url?: string) => any };
+declare let __dirname: string;
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/util/asyncLoad/system.ts:27]: works.
